extractor/image/daemon/image.go

```diff
@@ -0,0 +1,134 @@
+package daemon
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"golang.org/x/xerrors"
+)
+
+var mu sync.Mutex
+
+// image is a wrapper for github.com/google/go-containerregistry/pkg/v1/daemon.Image
+// daemon.Image loads the entire image into the memory at first,
+// but it doesn't need to load it if the information is already in the cache,
+// To avoid entire loading, this wrapper uses ImageInspectWithRaw and checks image ID and layer IDs.
+type image struct {
+	v1.Image
+	opener  opener
+	inspect types.ImageInspect
+}
+
+type opener func() (v1.Image, error)
+
+func Image(ref name.Reference) (v1.Image, func(), error) {
+	c, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("failed to initialize a docker client: %w", err)
```

Comment:
Do you think we should return an empty `cleanup` func instead of `nil`?

Reply:
Ah, you're right. It's better than nil.
https://github.com/aquasecurity/fanal/pull/96/commits/f3f7adc014ee85385d8fb0e8da047bb040e6c8fe